# Aging_aceleration

This set of codes was used to process public multiomic data and to estimate chronological and molecular DNA (mDNA) methylation clocks.

For TCGA data, TCGA TCGAbiolinks1 was used, which attempt to facilitate retrieval of GDC open-access data, for GEO datasets, GEOquery2 was used to retrieve raw, processed and clinical data. To process Illumina methylation data, the Sesame3 package waas used to preprocess, assess quality control, visualize and infer methylation data. 
To evaluate different public epigenetic clocks, the methylCIPHER, methylclock4 and eclock5 packages were computed.

REFERENCES
1.	Antonio Colaprico, Tiago Chedraoui Silva, Catharina Olsen, Luciano Garofano, Davide Garolini, Claudia Cava, Thais Sabedot, Tathiane Malta, Stefano M. Pagnotta, Isabella Castiglioni, Michele Ceccarelli, Gianluca Bontempi, Houtan Noushmehr. TCGAbiolinks. (Bioconductor, 2017). doi:10.18129/B9.BIOC.TCGABIOLINKS.
2.	gov>, S. D. <sdavis2@mail. GEOquery. (Bioconductor, 2017). doi:10.18129/B9.BIOC.GEOQUERY.
3.	Zhou, W., Triche, T. J., Jr, Laird, P. W. & Shen, H. SeSAMe: reducing artifactual detection of DNA methylation by Infinium BeadChips in genomic deletions. Nucleic Acids Res. 46, e123 (2018).
4.	Pelegí-Sisó, D., de Prado, P., Ronkainen, J., Bustamante, M. & González, J. R. methylclock: a Bioconductor package to estimate DNA methylation age. Bioinformatics 37, 1759–1760 (2021).
5.	Liu, Y. eClock: An ensemble-based method to accurately predict ages with a biased distribution from DNA methylation data. PLoS One 17, e0267349 (2022).
![image](https://user-images.githubusercontent.com/83257526/235824479-41cceaba-0ba9-4071-95be-27944945a8fc.png)



